master(engine): unexpected core dump for server master

### What did you do?

run a 1 servermaster 1 executor tiflow cluster 

### What did you expect to see?

run normally

### What did you see instead?

server master core dump
```
docker container logs 3c1f3f2a5ab2
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x1c17dfc]

goroutine 13341 [running]:
github.com/pingcap/tiflow/engine/pkg/notifier.(*Receiver[...]).Close(0x400199e090?)
	github.com/pingcap/tiflow/engine/pkg/notifier/notifier.go:65 +0x1c
github.com/pingcap/tiflow/engine/servermaster/serverutil.WatchExecutors({0x37aab80, 0x4001a5dd00}, {0xffff6fe33d78, 0x400032e380}, {0x37a8030, 0x4001a51fb0})
	github.com/pingcap/tiflow/engine/servermaster/serverutil/watch_executors.go:65 +0x3c4
github.com/pingcap/tiflow/engine/servermaster.(*Server).runLeaderService.func13()
	github.com/pingcap/tiflow/engine/servermaster/server.go:940 +0x9c
golang.org/x/sync/errgroup.(*Group).Go.func1()
	golang.org/x/sync@v0.0.0-20220722155255-886fb9371eb4/errgroup/errgroup.go:75 +0x60
created by golang.org/x/sync/errgroup.(*Group).Go
	golang.org/x/sync@v0.0.0-20220722155255-886fb9371eb4/errgroup/errgroup.go:72 +0xa8
```

log:
```
[2022/09/04 14:31:43.414 +00:00] [WARN] [discovery_keepalive.go:104] ["metastore session is done"] [executor-id=server-master-990e160a-c73b-4b4e-a0f8-d772d270ed53]
[2022/09/04 14:31:43.414 +00:00] [INFO] [gc_runner.go:67] ["default gc runner exited"]
[2022/09/04 14:31:43.414 +00:00] [WARN] [watch_executors.go:60] ["WatchExecutors took too long"] [duration=1m0.506749028s] [snap=null]
[2022/09/04 14:31:43.414 +00:00] [INFO] [watch_executors.go:52] ["WatchExecutors exited"] [error="watch executors: [DFLOW:ErrMasterSessionDone]master session is done"]
```
[server-master-0.log](https://github.com/pingcap/tiflow/files/9486542/server-master-0.log)

### Versions of the cluster

Dataflow Engine version (run `tiflow version`):

```console
master@12ff137
```

--- a/engine/servermaster/serverutil/watch_executors.go
+++ b/engine/servermaster/serverutil/watch_executors.go
@@ -54,7 +54,6 @@ func WatchExecutors(ctx context.Context, watcher executorWatcher, user executorI
 
 	watchStart := time.Now()
 	snap, updates, err := watcher.WatchExecutors(ctx)
-	defer updates.Close()
 
 	if duration := time.Since(watchStart); duration >= 100*time.Millisecond {
 		log.Warn("WatchExecutors took too long",
@@ -64,6 +63,7 @@ func WatchExecutors(ctx context.Context, watcher executorWatcher, user executorI
 	if err != nil {
 		return errors.Annotate(err, "watch executors")
 	}
+	defer updates.Close()
 
 	log.Info("update executor list", zap.Any("list", snap))
 	if err := user.UpdateExecutorList(snap); err != nil {

--- a/engine/servermaster/serverutil/watch_executors_test.go
+++ b/engine/servermaster/serverutil/watch_executors_test.go
@@ -19,13 +19,13 @@ import (
 	"testing"
 	"time"
 
-	"go.uber.org/atomic"
-
+	perrors "github.com/pingcap/errors"
 	"github.com/pingcap/tiflow/engine/model"
 	"github.com/pingcap/tiflow/engine/pkg/notifier"
 	"github.com/pingcap/tiflow/pkg/errors"
 	"github.com/stretchr/testify/mock"
 	"github.com/stretchr/testify/require"
+	"go.uber.org/atomic"
 )
 
 type mockExecutorWatcher struct {
@@ -147,6 +147,28 @@ func TestWatchExecutors(t *testing.T) {
 
 	cancel()
 	wg.Wait()
+}
+
+func TestWatchExecutorFailed(t *testing.T) {
+	t.Parallel()
+
+	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
+	defer cancel()
+
+	watcher := newMockExecutorWatcher()
+	user := newMockExecutorInfoUser()
+
+	evNotifier := notifier.NewNotifier[model.ExecutorStatusChange]()
+	defer evNotifier.Close()
+
+	watcher.On("WatchExecutors", mock.Anything).
+		Return(map[model.ExecutorID]string(nil),
+			(*notifier.Receiver[model.ExecutorStatusChange])(nil),
+			perrors.New("test error"),
+		).Times(1)
+
+	err := WatchExecutors(ctx, watcher, user)
+	require.ErrorContains(t, err, "test error")
 }
 
 func TestCloseWatchExecutors(t *testing.T) {

